vhost-device-spi/src/virtio_spi.rs

```diff
@@ -0,0 +1,71 @@
+// VirtIO SPI definitions
+//
+// Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+//          Haixu Cui <quic_haixcui@quicinc.com>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use bitflags::bitflags;
+use vmm_sys_util::{ioctl_ioc_nr, ioctl_ior_nr, ioctl_iow_nr};
+
+/// Linux SPI definitions
+/// IOCTL commands, refer Linux's Documentation/spi/spidev.rst for further details.
+const _IOC_SIZEBITS: u32 = 14;
+const _IOC_SIZESHIFT: u32 = 16;
+const SPI_IOC_MESSAGE_BASE: u32 = 0x40006b00;
+
+ioctl_ior_nr!(SPI_IOC_RD_BITS_PER_WORD, 107, 3, u8);
+ioctl_iow_nr!(SPI_IOC_WR_BITS_PER_WORD, 107, 3, u8);
+ioctl_ior_nr!(SPI_IOC_RD_MAX_SPEED_HZ, 107, 4, u32);
+ioctl_iow_nr!(SPI_IOC_WR_MAX_SPEED_HZ, 107, 4, u32);
+ioctl_ior_nr!(SPI_IOC_RD_MODE32, 107, 5, u32);
+ioctl_iow_nr!(SPI_IOC_WR_MODE32, 107, 5, u32);
+
+// Corresponds to the SPI_IOC_MESSAGE macro in Linux
+pub fn spi_ioc_message(n: u32) -> u64 {
+    let mut size: u32 = 0;
+    if n * 32 < (1 << _IOC_SIZEBITS) {
+        size = n * 32;
+    }
+    (SPI_IOC_MESSAGE_BASE | (size << _IOC_SIZESHIFT)) as u64
+}
+
+bitflags! {
+    pub struct LnxSpiMode: u32 {
+        const CPHA = 1 << 0;
+        const CPOL = 1 << 1;
+        const CS_HIGH = 1 << 2;
+        const LSB_FIRST = 1 << 3;
+        const LOOP = 1 << 5;
+        const TX_DUAL = 1 << 8;
+        const TX_QUAD = 1 << 9;
+        const TX_OCTAL = 1 << 13;
+        const RX_DUAL = 1 << 10;
+        const RX_QUAD = 1 << 11;
+        const RX_OCTAL = 1 << 14;
```

Comment:
I checked the spec text and this seems wrong. It says:

```
\field{mode_func_supported} indicates whether the following features are supported or not:
        bit 0-1: CPHA feature,
            0b00: invalid, must support as least one CPHA setting.
            0b01: supports CPHA=0 only;
            0b10: supports CPHA=1 only;
            0b11: supports CPHA=0 and CPHA=1;

        bit 2-3: CPOL feature,
            0b00: invalid, must support as least one CPOL setting.
            0b01: supports CPOL=0 only;
            0b10: supports CPOL=1 only;
            0b11: supports CPOL=0 and CPOL=1;

        bit 4: chipselect active high feature, 0 for unsupported and 1 for supported,
            chipselect active low must always be supported.

        bit 5: LSB first feature, 0 for unsupported and 1 for supported, MSB first must always be
            supported.

        bit 6: loopback mode feature, 0 for unsupported and 1 for supported, normal mode
            must always be supported.
```

Reply:
@epilys I was confused too, but IIUC these definitions comes from `include/uapi/linux/spi/spi.h`.

For that reason, I have already complained that the file name and the comment at the top of the file are misleading, because the definitions within it are not related to virtio at all.

Reply:
Sorry I didn't make it clear. 

`LnxSpiMode ` shows the SPI mode bit definitions, corresponds with the macros in `include/uapi/linux/spi/spi.h` 

While for bitmap of mode_func_supported in the spec, in virtio_spi.rs I use `ConfigMode`.

Reply:
I split virtio_spi.rs into two files, one for  virtio spi, another for linux spi bindings. I think this will make it more clear